lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) 2023, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import OverrideableBuilder from "supertokens-js-override";
+import { BaseRequest, BaseResponse } from "../../framework";
+import { NormalisedAppinfo } from "../../types";
+import { SessionContainer } from "../session";
+
+export type TypeInput = {
+    issuer: string;
+    defaultSkew?: number;
+    defaultPeriod?: number;
+    allowUnverifiedDevice?: boolean;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    issuer: string;
+    defaultSkew: number;
+    defaultPeriod: number;
+    allowUnverifiedDevices: boolean;
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    createDevice: (input: {
+        userId: string;
+        deviceName: string;
+        skew?: number;
+        period?: number;
+        userContext: any;
+    }) => Promise<{ status: "OK"; qr: string } | { status: "DEVICE_ALREADY_EXISTS_ERROR" }>;
```

Comment:
what is qr? a better variable name is needed

Reply:
qr is QR code info. Frontend should render this as a QR code.

Reply:
How about calling it `qrCode`?

Reply:
done.